libs/utils/analysis/frequency_analysis.py

```diff
@@ -163,6 +163,124 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
 # Plotting Methods
 ###############################################################################
 
+    def plotPeripheralClock(self, clk, title='Peripheral Frequency'):
+        """
+        Produce graph plotting the frequency of a particular peripheral clock
+
+        :param title: The title for the chart
+        :type  title: str
+
+        :param clk: The clk name to chart
+        :type  clk: str
+
+        :raises: KeyError
+        """
+        if not self._trace.hasEvents('clock_set_rate'):
+            self._log.warning('Events [clock_set_rate] not found, plot DISABLED!')
+            return
+        rate_df = self._dfg_trace_event('clock_set_rate')
+        enable_df = self._dfg_trace_event('clock_enable')
+        disable_df = self._dfg_trace_event('clock_disable')
+
+        rate_df = rate_df[rate_df.clk_name == clk]
+        rate_df['clock_setting'] = rate_df['rate']/1e6
+        rate_df['state'] = -1
+
+        pd.set_option('display.expand_frame_repr', False)
+
+        if not enable_df.empty:
+            enable_df = enable_df[enable_df.clk_name == clk]
+            enable_df['clock_setting'] = 1;
+        if not disable_df.empty:
+            disable_df = disable_df[disable_df.clk_name == clk]
+            disable_df['clock_setting'] = 0;
+
+        freq = pd.concat([rate_df, enable_df, disable_df])
+        freq.sort_index(inplace=True)
+
+        if freq.empty:
+            self._log.warning('No events for clock ' + clk + ' found in trace')
+            return
+
+        # This will extend the last frequency value to the end of the timerange.
+        last = freq.tail(n=1)
+        last.index = { self._trace.x_max }
+        freq = pd.concat([freq, last])
+        freq['timestamp'] = freq.index
+        freq['frequency'] = 0;
+
+        max_freq = rate_df['clock_setting'].max()
+        last_state = 1
+        last_freq = rate_df.iloc[0]['clock_setting']
+        for index, row in freq.iterrows():
```

Comment:
Had a look at the trace example in trappy, makes more sense now (I didn't know how the dataframes would be built).

It is indeed a tricky one. I couldn't think of a decent `apply()` use, but I think we can make it shorter and easier to read:

```
max_freq = rate_df['clock_setting'].max()
for index, row in freq.iterrows():
    # Get latest frequency update
    curr_freq = rate_df[:index]['clock_setting'].values[-1]
    freq.loc[index, 'frequency'] = 0 if not row.state else curr_freq
```

What do you think ?

Reply:
sure, adopted. Thanks!